### ISSUE TITLE
Documentation bug with VOLUME instruction

### ISSUE THREAD

**Issue by kuc** [docker/docker#14122]
`Docker version 1.6.2, build 7c8fca2`

From documentation:
https://docs.docker.com/reference/builder/#volume
>The `docker run` command initializes the newly created volume with any data that exists at the specified location within the base image.

(1) This works as excepted:
```
RUN mkdir -p /foo/bar && touch /foo/bar/test.txt
VOLUME /foo/bar
```

(2) but this not:
```
VOLUME /foo/bar
RUN touch /foo/bar/test.txt
```

File in image is created successfully, but when container is created `/foo/bar` becomes empty.
Strangely, (2) seems to work when `VOLUME` part is taken from cache, and `RUN` part is freshly appended to Dockerfile. But full rebuild of image breaks this.

------------

BUG REPORT INFORMATION

Use the commands below to provide key information from your environment:

docker version:
```
docker@boot2docker:~$ docker version
Client version: 1.6.2
Client API version: 1.18
Go version (client): go1.4.2
Git commit (client): 7c8fca2
OS/Arch (client): linux/amd64
Server version: 1.6.2
Server API version: 1.18
Go version (server): go1.4.2
Git commit (server): 7c8fca2
OS/Arch (server): linux/amd64
```

docker info:
```
docker@boot2docker:~$ docker info
Containers: 13
Images: 195
Storage Driver: aufs
 Root Dir: /mnt/sda1/var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 221
 Dirperm1 Supported: true
Execution Driver: native-0.2
Kernel Version: 4.0.3-boot2docker
Operating System: Boot2Docker 1.6.2 (TCL 5.4); master : 4534e65 - Wed May 13 21:24:28 UTC 2015
CPUs: 2
Total Memory: 1.957 GiB
Name: boot2docker
ID: EDZ6:D6NA:EKMP:HL4L:LUQ3:MGLZ:CAEW:EXEZ:HPJU:Q7RT:3CL3:TBNI
Debug mode (server): true
Debug mode (client): false
Fds: 54
Goroutines: 88
System Time: Sat May 30 08:05:47 UTC 2015
EventsListeners: 1
Init SHA1: 7f9c6798b022e64f04d2aff8c75cbf38a2779493
Init Path: /usr/local/bin/docker
Docker Root Dir: /mnt/sda1/var/lib/docker
```

uname -a:
```
docker@boot2docker:~$ uname -a
Linux boot2docker 4.0.3-boot2docker #1 SMP Wed May 13 20:54:49 UTC 2015 x86_64 GNU/Linux
```

Provide additional environment details (AWS, VirtualBox, physical, etc.):
- VirtualBox, boot2docker

List the steps to reproduce the issue:
Describe the results you received:
Describe the results you expected:
See above.

----------END REPORT ---------

#ENEEDMOREINFO

**Comment by phemmer** [comment]
Just a note, this behavior is documented, just on a different page.

https://docs.docker.com/userguide/dockervolumes/

> Changes to a data volume will not be included when you update an image.

See also #13117

Perhaps this line should be copied into the builder reference doc.

**Comment by kuc** [comment]
>Changes to a data volume will not be included when you update an image.

It's still not telling that after declaring volume you can't add any files to its directory when creating image, and that this should be done only before `VOLUME` instruction.
Clarification in documentation will be very appreciated.

**Comment by phemmer** [comment]
> It's still not telling that after declaring volume you can't add any files to its directory when creating image

That's *exactly* what it's saying.

**Comment by kuc** [comment]
Not clear enough for me, sorry. I'm not "updating image", I'm running `docker build` on `Dockerfile` and files are missing because instruction order matters.

Paradoxically, "updating image" works for me:
>(2) seems to work when VOLUME part is taken from cache, and RUN part is freshly appended to Dockerfile. But full rebuild of image breaks this.

**Comment by cpuguy83** [comment]
@kuc Do you have some spare time to update the relevant docs?

**Comment by kuc** [comment]
@cpuguy83 Here: https://github.com/docker/docker/blob/master/docs/reference/builder.md#volume ?
I could try to find time to do this in upcoming days...

**Comment by cpuguy83** [comment]
Closing as this was addressed by #15337
